runtimes/python-3.10/server.py

```diff
@@ -47,58 +47,95 @@ def redirect(self, url, statusCode = 301, headers = {}):
         return self.send('', statusCode, headers)
 
 class Request:
-    rawBody = None
+    bodyString = None
     body = None
     headers = None
     method = None
     url = None
+    path = None
+    port = None
+    query = None
+    queryString = None
+    scheme = None
+    host = None
 
 class Context:
     req = Request()
     res = Response()
 
-    _logs = []
-    _errors = []
+    logs = []
+    errors = []
 
     def __init__(self):
-        self._logs = []
-        self._errors = []
+        self.logs = []
+        self.errors = []
         self.req = Request()
         self.res = Response()
 
-    # TODO: Support for infinite parameters
-    # TODO: Support for objects (stringify)
     def log(self, message):
-        self._logs.append(str(message))
+        if isinstance(message, (list, dict, tuple)):
+            self.logs.append(json.dumps(message, separators=(',', ':')))
+        else:
+            self.logs.append(str(message))
 
     def error(self, message):
-        self._errors.append(str(message))
+        if isinstance(message, (list, dict, tuple)):
+            self.errors.append(json.dumps(message, separators=(',', ':')))
+        else:
+            self.errors.append(str(message))
 
 HTTP_METHODS = ['GET', 'HEAD', 'POST', 'PUT', 'DELETE', 'CONNECT', 'OPTIONS', 'TRACE', 'PATCH']
 
 @app.route('/', defaults={'u_path': ''}, methods = HTTP_METHODS)
 @app.route('/<path:u_path>', methods = HTTP_METHODS)
 def handler(u_path):
-    if (request.headers.get('x-open-runtimes-secret') != os.getenv('OPEN_RUNTIMES_SECRET')):
+    if (request.headers.get('x-open-runtimes-secret', '') != os.getenv('OPEN_RUNTIMES_SECRET', '')):
         return 'Unauthorized. Provide correct "x-open-runtimes-secret" header.', 500
 
     context = Context()
 
-    url = urlparse(request.url)
-    path = url.path
-    query = url.query
-    if query:
-        path += '?' + query
-
-    context.req.rawBody = request.get_data(as_text=True)
-    context.req.body = context.req.rawBody
+    context.req.bodyString = request.get_data(as_text=True)
+    context.req.body = context.req.bodyString
     context.req.method = request.method
-    context.req.url = path
     context.req.headers = {}
 
+    context.req.path = request.path
+    context.req.scheme = request.headers.get('x-forwarded-proto', 'http')
+
+    url = urlparse(request.url)
+    context.req.queryString = url.query or ''
+    context.req.query = {}
+
+    for param in context.req.queryString.split('&'):
+        pair = param.split('=')
+
+        if pair[0]:
+            context.req.query[pair[0]] = pair[1]
+
+    host = request.headers.get('host', '')
+    if ':' in host:
+        context.req.host = host.split(':')[0]
+        context.req.port = int(host.split(':')[1])
```

Comment:
why do you need to use string splitting when you have the parsed URL already ?

Reply:
host header could be a domain like `myapp.com`, or it could be IP with port like `12.44.245.12:3000`. We split to separate port and host from this header.

Reply:
No I was not referring specifically to the host & port. I was referring to the manual parsing of strings as opposed to using the standard library

Reply:
Ill use std library.